consensus/src/experimental/buffer_manager.rs

```diff
@@ -254,6 +184,112 @@ impl StateManager {
+                let commit_ledger_info = LedgerInfo::new(
+                    executed_blocks.last().unwrap().block_info(),
+                    ordered_box
+                        .ordered_proof
+                        .ledger_info()
+                        .consensus_data_hash(),
+                );
+
+                self.signing_phase_tx
+                    .send(SigningRequest {
+                        ordered_ledger_info: ordered_box.ordered_proof.clone(),
+                        commit_ledger_info,
+                    })
+                    .await?;
+
+                set_elem(
+                    &current_cursor,
+                    buffer_item.advance_to_executed(executed_blocks),
+                );
+                self.execution_root = current_cursor;
+            } else {
+                // even if there is a sync happened before the response
+                // the buffer item right after execution root should be an ordered buffer item
+                return Err(anyhow!("Inconsistent buffer item state"));
+            }
+        }
+        Ok(())
+    }
+
+    /// this function handles the execution response and updates the buffer
+    /// if the execution fails: it re-collects a larger batch and retries.
+    async fn process_execution_resp(
+        &mut self,
+        execution_resp: ExecutionResponse,
+    ) -> anyhow::Result<()> {
+        // we do not use callback from the execution phase to fetch the retry blocks
+        // because we want the buffer accessed by a single thread
+        let res = self.update_buffer_execution_resp(execution_resp).await;
+        if res.is_err() {
+            // it might be possible that the buffer is already reset
+            // in which case we are retrying an irrelevant large batch
+            // this is ok as blocks can be executed more than once
+            let mut cursor = self.buffer.head.clone();
+            let mut ordered_blocks: Vec<ExecutedBlock> = vec![];
+            while cursor.is_some() {
+                ordered_blocks.extend(get_elem(&cursor).get_blocks().clone());
+                if link_eq(&cursor, &self.execution_root) {
+                    // there must be a successor since the last execution failed
+                    cursor = get_next(&cursor);
+                    ordered_blocks.extend(get_elem(&cursor).get_blocks().clone());
+                    // retry execution with the larger batch
+                    // send blocks to execution phase
+                    self.execution_phase_tx
+                        .send(ExecutionRequest { ordered_blocks })
+                        .await?;
+                    break;
+                }
+                cursor = get_next(&cursor);
+            }
+            // the only case that cursor did not meet execution root is when the buffer is empty
+            // in which case we do nothing
+        } else {
+            // try the next item
+            let cursor = get_next(&self.execution_root);
```

Comment:
I think this should be outside of process_response function, maybe in an orchestrated loop (later PR)

Reply:
Noted on todo